### PR TITLE
Vs2015 support

### PR DESCRIPTION
**Pull request by KevinRansom**
Fixes 182

Ensure that the VisualFsharp projects build when installed on a PC with VS2015 installed and no VS2013.

Fixup references from VS2013 VS assemblies to VS 2015 assemblies.  Also the prototype compiler build had bit-rot a little.

Kevin

### CODE REVIEW

**Comment by dsyme** [comment]
Hi Kevin.

As far as I understand it, this would mean the compiler and core library no longer compile with VS2013 and a VS2015 preview install would be required to work on the language and compiler.

I don't think we can do that. It's one thing for the Visual F# IDE Tools to require VS2015, but the compiler and library itself need to compilable with the currently available Visual Studio vNext-1 RTM - i.e. Visual Studio 2013.

Please let's not require VS2015 preview installs.  If we diid then I estimate engagement rate will drop to around 10% of where it is now - we will lose 9/10 contributors. Even I will find it difficult, as I do nearly all my language and compiler work from my laptop where I do not have VS2015 preview installed (for stability reasons).  

If anything, the repo needs to be moving in the other direction and enabling work on the VF# IDE Tools using VS2013 (and also enabling contributions to the compiler using Linux/Mono)  I know there was a thread on that, and you made a decision not to allow testing using VS2015. But equally the contribution rate to the VF#IDE Tools is lower. I think that's partly because VS2015 preview installs are required. I know I've found it difficult to contribute to them because of that.

This is also why AppVeyor is failing. AppVeyor is like a typical contributor, it has VS2013 installed.

**Comment by forki** [comment]
>  compiler and library itself need to compilable with the currently available Visual Studio

why do we need VS for that?

**Comment by dsyme** [comment]
> why do we need VS for that?

Agreed, I meant "VS2013 or it's corresponding SDKs".  It looks like VS2015 comes with an updated MSBuild SDK.

**Comment by KevinRansom** [comment]
The compiler takes a dependency on MsBuild, the OSS Codebase includes the complete set of Visual Studio Addins for F#.  The current CI solution does not build or test the VS addins and that is a problem becaue the compiler and library changes can still break the VS Addins.  Just doing a CI on the compiler and libraries, because they are the "Fun Stuff" is not acceptable.

I understand that the requirement to have VS 2015 installed in order to build and verify changes to F# is a burden, although working at MS, as I do, I am very used to installing the latest builds of anything I am working on and so the community hesitation about moving forward seems strange to me, although I suppose I can see the point intellectually.

When thinking about a programing language it is easy to care most about the compiler, however it is important to recognize is that the Visual F# Tools consists of _**the compiler, F# core, FSI, and the VS addin components.**_  Also the overwhelmingly most common use of F# is from the latest release of Visual Studio.  The most common feedback we get is "Why does the F# tooling not support X or Y editor feature?" that is in VB and C#.

For the purposes of this repo the CI needs to work with VS2015 since that is the deployment vehicle and host for this release of Visual F# Tools.

The original port to VS2015 assumed that VS2015 was the build environment and the expected target environment.  Our internal build does the correct thing, however, the port of the open source build to VS 2015 turned into a bit of a mutant with dependencies on both VS 2013, VS 2015, although it probably works with F# 3.1.2 Oob and VS 2015 installed (I haven't tested it).

I will see what I can do to make the build for the entire stack work with VS2013.  Although the CI will be based on VS2015 and will eventually compile and test the VS addins.

I urge everyone to dogfood the VS2015 bits, because that is what everyone is going to be using to use F# 4.0 and we need to find and fix whatever bugs we can before we ship.

**Comment by dungpa** [comment]
Hi Kevin,

Is this possible to make the compiler solution compatible with VS 2013 and the VS integration solution compatible with VS 2015?

In that way, it won't disrupt current workflows that contributors have been used to.

I often use VS 2015 when I play with the VS integration part. However, when working on the compiler and core library, I prefer VS 2013 since it's more stable.

**Comment by KevinRansom** [comment]
Hey Mate,

Visual F# Tools ships with VS2015, we will build and test internally against the VS2015 toolset.  I will do my best to make it build against the VS2013 toolset as a convenience to folk who are unwilling or unable to install VS 2015.  I hope to be able to make it test against VS2013 too, but that is not a commitment I am prepared to make right now.  Using different compilers for different parts of the product is not that interesting. I will require the CI to run with the latest VS 2015 preview and compile and test VS addins because that is what we will deploy.  It is intended to catch bugs when a PR builds and runs on VS2013 and fails on VS2015.  I don't expect that to be super common, but it is something the community will need to guard against.  Certainly that describes the current state of the OSS build codebase.

Once it's done, I doubt if most people will even notice.  However, I am quite disappointed that it is believed to be necessary.

Kevin

**Comment by forki** [comment]
Yes the CI should build/test everything. That's super important - especially since the full test suite is very very slow.

------------------------------------------------------------------------------------------------

Regarding VS: I really care about F# as a language and love working with you guys. It's absolutly amazing to be part of this. I want to see F# work everywhere (which explicitly includes VS 2015) and I spent a lot of my free time to help to make this happen (on a lot of of OSS projects).

But my main personal interest is still F# the language - not VS the product.

From the mission statement at http://foundation.fsharp.org:

    "The FSSF acknowledge the role of Microsoft as the current primary developers of the F# language, open-source compiler and open-source core tools, with a focus on the Windows platform. They accept that as an excellent ongoing arrangement. They welcome Microsoft’s contributions to the aims of the FSSF through both the open source releases of the F# compiler and the commercial releases of the Visual F# tooling."

This statement and a general agreement in the community means language improvements go into this repo here and not directly into https://github.com/fsharp/fsharp. Whatever gets accepted here, is what the language looks like. Otherwise we would basically fork the language and we obviously don't want this.

But there are also implications of this statement which are slightly in contrast with Microsoft's primary goal. If we want to make this repo the one and only truth for F# the language, then the language (compiler + fsi) should build basically everywhere. People who don't have a Microsoft background should still be able to work on the language, it's compiler and the fsi.

That said I think both is possible ;-)

**Comment by dsyme** [comment]
One thing I've taken away from this is that, as someone who cares about the Visual F# Tools in Visual Studio 2015, I'll definitely be installing the Visual Studio 2015 CTP5 when I get back online, so I can help develop and use the product in that configuration.  As we near a Visual Studio release, it's important for many of us (but not necessarily all) to be doing that.

I think it's really great that the CI will soon use the Visual Studio 2015 Preview, and that the CI will be on a path towards being able to include building and unit-testing the Visual F# Tools.  Fantastic!

One way to loko at this is that this repo is about two things

* First, and most importantly for Microsoft and all users of Visual Studio, it's about shipping a wonderful and important product - right now that means the VIsual F# Tools in Visual Studio 2015.  

* Second, the F# community are happy that this repo is the path for contributions to the F# Compiler and Core Library, because of the unity, integrity, and quality QA value this brings.  

This dual purpose of the repo occasionally brings up questions like the above, especially as we move close to a new release of Visual Studio. At times like this, there is naturally a very strong need to get people to move on to the new version of Visual Studio for testing and quality purposes.

**Comment by KevinRansom** [comment]
Whilst our planning is very far from complete for V.Next, cross platform will be one of the main themes.  You may have noticed that I have started a branch in my fork related to the netcore clr which is likely to be the main cross platform play for .NET according to the November announcements.  That was not a theme for this release of Visual F#, even so we enabled new features that are headed for mono and even encouraged development of shebang support which is uninteresting in windows scenarios, but useful in UNIX scripting scenarios.  I think our cross plat credibility is adequate given how recently we adopted open development.  The Main issue with focusing on VS is that it is a proper shipping project.  The discipline necessary to ship a high quality, broad market product is immense.  We will not compromise that quality in order to enable hobby project scenarios. Also  it is not our position that the Visual FSharp Tools is the F# compiler ➕  some VS addins.  To us it is a stack including compiler and tooling in the same way that Roslyn is the compilers ➕ tooling.  Seeking to separate the two projects even in your own minds will undoubtedly lead to unsatisfactory outcomes to everybody.  I have heard the feedback and will do what I can to enable you to continue to use VS 2013 and its SDK as a convenience feature of the build.

**Comment by dungpa** [comment]
I think you misunderstood our intentions. I agree that the product is the integrated stack of compilers and tooling. I have contributed into VS integration part and would like to continue doing so. Contributions have been sparse in that area...

Our suggestion is to keep the compiler solution usable in VS 2013. It would keep the bar for contribution as low as it is now. I have VS 2015 (+ SDK) on all dev machines; whether or not the solutions can be built in VS 2013 doesn't matter much to me. But I think it makes differences to potential contributors.

**Comment by KevinRansom** [comment]
I will do what I can to make 2013 a usable development tool for the entire stack.  I expect to be able to get quite close.

**Comment by enricosada** [comment]
:+1: for vs 2013 compatibility until rtm, thx @KevinRansom

**Comment by KevinRansom** [comment]
Appveyor accepted these changes with both VS 2013 and VS2015

Let me know what issues you experience when building.

Please note VS selection is automagic.  
* If you open a VS2013 command window it will use VS2013
* if you open a VS2014 command window it will use VS2015

File: src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj [hunk 1]
```diff
@@ -15,9 +15,9 @@
     <FileAlignment>512</FileAlignment>
   </PropertyGroup>
   <ItemGroup>
-      <FsSrGen Include="..\FSInteractiveSettings.txt">
```
Comment (latkin) [inline src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj]: Changes to this file are just whitespace, it can be removed from the PR

File: src/FSharpSource.msbuildreferences.targets [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (latkin) [inline src/FSharpSource.msbuildreferences.targets]: I don't understand the need to add new files? We already have 2 targets files.

File: src/FSharpSource.msbuildreferences.targets [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+<!-- Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!-- F# compiler has a depedency on MsBuild.  Different VS and F# Deployments carry various MsBuild versions  -->
+  <!-- This section sets the MsBuild references the compiler requires depending on the value of the environment variable  -->
+  <!-- VisualStudioVersion, 12.0 for VS2013, 1 for VS2015 -->
+  <!-- By Default uses VS2013 values -->
+  <Choose>
+    <When Condition="'$(VisualStudioVersion)' == '14.0'">
+      <PropertyGroup>
```
Comment (latkin) [inline src/FSharpSource.msbuildreferences.targets]: It would be much cleaner to define simply the version (either 12.0 or 14.0) in a common targets file, then consumers use that version. Something like:
```
<Reference Include="Microsoft.Build.Utilities.core, Version=$(VSVersion)"/>
```

Redefining everything multiple times is not needed, and we should avoid introducing 6 new build variables.

File: src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj [hunk 4]
```diff
@@ -47,11 +47,11 @@
     <Reference Include="$(FSCoreLKGPath)" />
     <Reference Include="System" />
     <Reference Include="System.Numerics" />
-    <Reference Include="Microsoft.Build.Framework, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
-    <Reference Include="Microsoft.Build.Engine, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
-    <Reference Include="Microsoft.Build, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
-    <Reference Include="Microsoft.Build.Utilities.v12.0" />
-    <Reference Include="Microsoft.Build.Tasks.v12.0" />
+    <Reference Include="Microsoft.Build.Framework, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
```
Comment (latkin) [inline src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj]: Why is proto using VS 2010-era msbuild version, but everything else is using VS 2015 version?

File: tests/RunTests.cmd [hunk 5]
```diff
@@ -79,7 +81,7 @@ if /I "%2" == "ideunit" (goto :IDEUNIT)
 
 echo Usage:
 echo.
-echo RunTests.cmd ^<debug^|release^> ^<fsharp^|fsharpqa^|coreunit^|coreunitportable47^|coreunitportable7^|coreunitportable78^|coreunit259^|ideunit^|compilerunit^> [TagToRun^|"Tags,To,Run"] [TagNotToRun^|"Tags,Not,To,Run"]
+echo RunTests.cmd ^<debug^|release^|vsdebug^|vsrelease^> ^<fsharp^|fsharpqa^|coreunit^|coreunitportable47^|coreunitportable7^|coreunitportable78^|coreunit259^|ideunit^|compilerunit^> [TagToRun^|"Tags,To,Run"] [TagNotToRun^|"Tags,Not,To,Run"]
```
Comment (latkin) [inline tests/RunTests.cmd]: New test configurations are needed to unblock build? What are these configs doing that debug and release can't do?

File: vsintegration/src/fsharp.vsreferences.targets [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (latkin) [inline vsintegration/src/fsharp.vsreferences.targets]: Again, let's please avoid needlessly adding more and more files. We have targets files already that all projects import, why can't we use those?

File: vsintegration/src/unittests/Unittests.fsproj [hunk 7]
```diff
@@ -3,14 +3,13 @@
...
     <FSharpSourcesRoot>..\..\..\src</FSharpSourcesRoot>
-    <vsFrameworkLatest>Microsoft.VisualStudio.Shell.12.0</vsFrameworkLatest>
     <ProjectLanguage>FSharp</ProjectLanguage>
     <SIGN_WITH_MSFT_KEY>true</SIGN_WITH_MSFT_KEY>
   </PropertyGroup>
   <Import Project="$(FSharpSourcesRoot)\FSharpSource.settings.targets" />
+  <Import Project="$(ProjectDirectory)..\fsharp.vsreferences.targets" />
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform>x86</Platform>
```
Comment (latkin) [inline vsintegration/src/unittests/Unittests.fsproj]: Why the change from `Platform` to `PlatformTarget`?

File: appveyor.yml [hunk 8]
```diff
@@ -1,6 +1,14 @@
+os: Visual Studio 2015 CTP
+
 init: 
 build_script: 
   - cmd: appveyor-build.cmd
+
+# scripts that run after cloning repository
+install:
+  # by default, all script lines are interpreted as batch
+  - set PATH=C:\Program Files (x86)\MSBuild\14.0\Bin;%PATH%
```
Comment (latkin) [inline appveyor.yml]: `C:\` drive should not be hardcoded.
`Program Files (x86)` should not be hardcoded.

**Comment by latkin** [comment]
Hmm, build now fails for me with this applied
```
"C:\Users\latkin\Source\Repos\visualfsharp\vsintegration\fsharp-vsintegration-build.proj" (default target) (1) ->
"C:\Users\latkin\Source\Repos\visualfsharp\vsintegration\src\Deployment.sln" (Build target) (2) ->
"C:\Users\latkin\Source\Repos\visualfsharp\vsintegration\src\deployment\EnableOpenSource\EnableOpenSource.csproj" (defa
ult target) (3) ->
"C:\Users\latkin\Source\Repos\visualfsharp\vsintegration\src\vs\FsPkgs\FSharp.Project\Common.Source.CSharp\Project\Proj
ectSystem.Base.csproj" (default target) (12) ->
(CoreCompile target) ->
  ProjectFactory.cs(367,60): error CS0234: The type or namespace name 'Conversion' does not exist in the namespace 'Mic
rosoft.Build' (are you missing an assembly reference?) [C:\Users\latkin\Source\Repos\visualfsharp\vsintegration\src\vs\
FsPkgs\FSharp.Project\Common.Source.CSharp\Project\ProjectSystem.Base.csproj]
  ProjectFactory.cs(771,56): error CS0234: The type or namespace name 'Conversion' does not exist in the namespace 'Mic
rosoft.Build' (are you missing an assembly reference?) [C:\Users\latkin\Source\Repos\visualfsharp\vsintegration\src\vs\
FsPkgs\FSharp.Project\Common.Source.CSharp\Project\ProjectSystem.Base.csproj]
```

I don't use a VS command prompt, could that be the problem?

File: src/fsharp/ReferenceResolution.fs [hunk 9]
```diff
@@ -106,7 +112,10 @@ module internal MSBuildResolver =
             | Net35 ->  Some TargetDotNetFrameworkVersion.Version35
             | Net40 ->  Some TargetDotNetFrameworkVersion.Version40
             | Net45 ->  Some TargetDotNetFrameworkVersion.Version45
+#if BUILDING_PROTO
+#else
             | Net451 -> Some TargetDotNetFrameworkVersion.Version451
```
Comment (latkin) [inline src/fsharp/ReferenceResolution.fs]: We should not need to make these changes... `Net451` is a valid enum value as of msbuild 12.0 (docs [here](https://msdn.microsoft.com/en-us/library/ms126273(v=vs.121).aspx)).  I believe this is only causing a problem because the proto msbuild version was changes down to 4.0 from 12.0.

File: src/fsharp/ast.fs [hunk 10]
```diff
@@ -1742,7 +1755,7 @@ let mkSynDotBrackSeqSliceGet  m mDot arr (argslist:list<SynIndexerArg>) =
                        | SynIndexerArg.One x -> yield x
                        | _ -> () ]
     if notsliced.Length = argslist.Length then
-        SynExpr.DotIndexedGet(arr,[SynIndexerArg.One (SynExpr.Tuple(notsliced,[],unionRanges (Seq.head notsliced).Range (Seq.last notsliced).Range))],mDot,m)
```
Comment (latkin) [inline src/fsharp/ast.fs]: The problem here is that although we create a proto version of FSharp.Build.dll, for some reason we don't create a proto version of FSharp.Core. This should probably be corrected.

The internal official build does create a proto version of FSharp.Core, and references that when compiling proto ver of FSharp.Compiler.  Right now OSS is just referencing the LKG FSharp.Core here, which is missing a lot of stuff.

File: src/FSharpSource.msbuildreferences.targets (hunk 3 above)
Comment (KevinRansom) [inline src/FSharpSource.msbuildreferences.targets]: This is actually quite convenient and means we can move things forward if necessary by editing one of these two files.  Unfortunately the assembly names change for some of the assemblies as well as the version numbers.  I think this is reasonably convenient.

I would prefer not to have added two additional target files.  However, the existing two files are not used uniformly by all of the projects.  Perhaps we can fix that in the future.

File: src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj (hunk 4 above)
Comment (KevinRansom) [inline src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj]: Proto build is built by the LKG, the assembly resolution mechanism used by the LKG objects to assemblies that do not belong to the target framework.  The Dev12 msbuild files were compatible with net45, the Dev 14 ones are compatible with net451.

File: src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj (hunk 1 above)
Comment (KevinRansom) [inline src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj]: I prefer to tidy it up.

File: src/fsharp/ReferenceResolution.fs (hunk 9 above)
Comment (KevinRansom) [inline src/fsharp/ReferenceResolution.fs]: The current OSS build incorrectly sets the FX_ATLEAST_45 variable, if we ensure the build doesn't set this variable then we don't need these additional #defines.  However now we have warnings as errors we will need to ensure the unused parameters don't cause the warnings.

File: src/fsharp/ast.fs (hunk 10 above)
Comment (KevinRansom) [inline src/fsharp/ast.fs]: I agree we should make a proto fsharp.core build.  I will add an issue, it is beyond the scope of this checkin, which is already 10 times bigger than I expected :-)

File: tests/RunTests.cmd (hunk 5 above)
Comment (KevinRansom) [inline tests/RunTests.cmd]: These build configurations have been present since the ui stacks were open sourced.  They differ from release and debug in that they use delay signed Microsoft keys.  We need to sign with the Microsoft keys to make the VS integration work.  We could eliminate the test keys, I'm not sure how that will impact existing the workflow of F# open source developers.

File: src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj (hunk 4 above)
Comment (dsyme) [inline src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj]: Hmm... I still don't really understand why v4.0 msbuild is being used.  For example, AFAIK Mono doesn't even have v4.0 anymore.  Could you explain a bit more why not use 12.0 here?

File: src/fsharp/ast.fs (hunk 10 above)
Comment (dsyme) [inline src/fsharp/ast.fs]: Change the code in the compiler to use List.head and List.last since we define List.last in src\absil\illib.fs in all situations. (If you do add something like Seq.last to the compiler, add it in src\absil\illib.fs)

File: vsintegration/src/fsharp.vsreferences.targets (hunk 6 above)
Comment (dsyme) [inline vsintegration/src/fsharp.vsreferences.targets]: @latkin  - I don't think there is a shared imported targets file for "vsintegration" - only for the compiler/library - so I think adding this file is ok and the best way to factor the swicth between VS references.

File: vsintegration/src/unittests/Unittests.fsproj (hunk 7 above)
Comment (dsyme) [inline vsintegration/src/unittests/Unittests.fsproj]: It does look like "PlatformTarget" is the right thing to use http://stackoverflow.com/questions/1074654/how-do-i-force-msbuild-to-compile-for-32-bit-mode

File: src/FSharpSource.msbuildreferences.targets (hunk 2 above)
Comment (dsyme) [inline src/FSharpSource.msbuildreferences.targets]: I do think this could be added to the end of FSharpSource.targets, but I don't mind too much either way

File: src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj (hunk 4 above)
Comment (dsyme) [inline src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj]: For example, here is the log from building from a VS2013 command prompt on my machine (have VS2010, VS2012, VS2013, VS2015 installed) https://gist.github.com/dsyme/b3838c25b2285bdff0aa

At this line when building FSHarp.Build-proto.dll the MSBuild v4.0 files are being referenced from the GAC.  Ths doesn't seem right https://gist.github.com/dsyme/b3838c25b2285bdff0aa#file-gistfile1-txt-L464

It also seems to happen later when building FSharp.Build.dll at this line: https://gist.github.com/dsyme/b3838c25b2285bdff0aa#file-gistfile1-txt-L574

File: src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj (hunk 1 above)
Comment (latkin) [inline src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj]: Per contribution guidelines, pure cleanup should be done separately. This just adds needlessly to the diff.

> Code cleanup which is unrelated to a bug fix or feature should generally be made separate to other checkins where possible.

File: vsintegration/src/fsharp.vsreferences.targets (hunk 6 above)
Comment (latkin) [inline vsintegration/src/fsharp.vsreferences.targets]: Most of the vsintegration stuff already imports FSharpSource.targets. It would be ideal if we could keep targets/settings like this together.

File: src/fsharp/ast.fs (hunk 10 above)
Comment (latkin) [inline src/fsharp/ast.fs]: Honestly I don't care much which things get proto-built. This is just one more example of how the OSS build diverges from the Microsoft build, which causes tremendous confusion and complexity for me and @KevinRansom.

Just to get it down "on paper" while I'm thinking about it, the two biggest pain points are (indirectly related to this PR):

- Proto
  - OSS build uses proto FSharp.Compiler, fsc, **FSharp.Build**
  - Msft build uses proto FSharp.Compiler, fsc, **FSharp.Core**
- LKG
  - OSS will use RTM 3.0 or 3.1 F# compiler/targets as the LKG if it finds them in the regular place
  - OSS will use the *4.0.30319.1* LKG otherwise, and for powerpack tools
  - Msft build always uses LKG for both compiler and powerpack, version *2.0.50726.900*

Just these two primary differences cause frequent issues where stuff will build for OSS, but not internally, or vice versa, and it's not trivial to work through the resolution.  I don't think anyone is against simplifying/unifying, but the whole system is so fragile that we are wont to make sweeping changes for fear of silently retargeting something the wrong way.

There are also numerous simpler issues (different NUnit version for tests, different ways of defining InternalsVisibleTo and versions, different hierarchy of projects, slightly different sets of #defines, different set of target platforms, etc) which don't make it any easier.

And now this week we are trying to accommodate requests to make the whole thing work for both VS 2013 and VS 2015 at the same time.

Sorry for the rant. I know @KevinRansom is just trying to get everything working. I'm just really trying to avoid introducing yet more complexity and divergence in the build/test infra.

File: src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj (hunk 1 above)
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj]: I think we should be lenient on this.

File: appveyor.yml (hunk 8 above)
Comment (KevinRansom) [inline appveyor.yml]: I actually copied this from the appveyor docs, I wasn't really thinking.  I will fix

File: src/fsharp/ast.fs (hunk 10 above)
Comment (KevinRansom) [inline src/fsharp/ast.fs]: I agree with Lincoln about the need for a proto version of fsharp the Se.list thing below just shouldn't have happened if we were set up correctly from the beginning.

File: tests/RunTests.cmd (hunk 5 above)
Comment (KevinRansom) [inline tests/RunTests.cmd]: I will eliminate these in a later checkin, I need them for now.

File: vsintegration/src/unittests/Unittests.fsproj (hunk 7 above)
Comment (KevinRansom) [inline vsintegration/src/unittests/Unittests.fsproj]: Typo, will fix

File: src/fsharp/ast.fs (hunk 10 above)
Comment (dsyme) [inline src/fsharp/ast.fs]: TBH the Microsoft internal build has always been a mystery to me :) Agreed we should align the builds as much as possible though, certainly the LKGs should be the same.

FWIW I suspect we don't need to build either a proto FSharp.Core nor a proto FSharp.Build with the LKG.  I didn't know the Microsoft internal build made a proto FSharp.Core and I don't know why it would need to - the LKG comes with the FSharp.Core it like to use.   There's no real guarantee the LKG will even be suitable for building the latest FSharp.Core.

As a historical note, the original duplication of ther project files crept in when someone (no one who is still here :)) duplicated all the project files way back in 2007 to make the DevDiv build projects, for reasons I never qutie understood. We continued using the original project files too. Misalignments crept in each time we needed to do an open source code drop for F# 2.0-3.1 - we needed a build that worked in the open and couldn't extract the DevDiv build files. So we put together something that worked based on the original project files. 

Anyway we've been living with two build systems on and off ever since that original duplication in 2007. It's amazing how a decision to duplicate things propagates down the years.  Now we're finally open source we can align things, let's do it.

Cheers
Don
